interfaces/builtin/mount_control.go

```diff
@@ -368,21 +417,38 @@ func validateMountOptions(mountInfo *MountInfo) error {
 	if len(mountInfo.options) == 0 {
 		return errors.New(`mount-control "options" cannot be empty`)
 	}
-	// With "functionfs" none of the valid "options" can be harmful so no need to check against allowedMountOptions
-	if mountInfo.isType("functionfs") {
-		if err := validateNoAppArmorRegexpWithError(`cannot use mount-control "option" attribute`, mountInfo.options...); err != nil {
-			return err
+	if err := validateNoAppArmorRegexpWithError(`cannot use mount-control "option" attribute`, mountInfo.options...); err != nil {
+		return err
+	}
+	var types []string
+	if mountInfo.hasType() {
+		fsExclusiveOption := optionIncompatibleWithFsType(mountInfo.options)
+		if fsExclusiveOption != "" {
+			return fmt.Errorf(`mount-control option %q is incompatible with specifying filesystem type`, fsExclusiveOption)
 		}
+		types = mountInfo.types
 	} else {
-		for _, o := range mountInfo.options {
-			if !strutil.ListContains(allowedMountOptions, o) {
-				return fmt.Errorf(`mount-control option unrecognized or forbidden: %q`, o)
+		types = defaultFSTypes
+	}
+	for _, o := range mountInfo.options {
+		if strutil.ListContains(allowedKernelMountOptions, o) {
+			continue
+		}
+		if strutil.ListContains(allowedUserspaceMountOptions, o) {
+			continue
+		}
+		foundAllowed := false
+		optionName := strings.Split(o, "=")[0]
+		for _, fstype := range types {
+			fsAllowedOptions, exists := allowedFilesystemSpecificMountOptions[fstype]
+			if exists && strutil.ListContains(fsAllowedOptions, optionName) {
+				foundAllowed = true
+				break
 			}
```

Comment:
maybe I'm  misunderstanding the context, but shouldn't the option in this context be allowed by all the filesystems and not just one of them?

Reply:
That's a good point. We probably want to restrict allowed fs-specific options to only those supported by all filesystem types listed by the user, rather than any one of them. Otherwise, an option which is illegal for one FS type could be used by another simply by listing the other in the FS `type:` list. I'll make the change.